File: go/internal/utils/security.go
package utils

import (
	"fmt"
	"go/ast"
	"go/parser"
	"go/token"
	"io/ioutil"
	"os"
	"regexp"
	"runtime"
	"strings"
	"sync"
	"time"
)

// SecurityChecker 安全检查器
type SecurityChecker struct {
	sensitivePatterns map[string][]string
	mu               sync.RWMutex
}

// NewSecurityChecker 创建新的安全检查器
func NewSecurityChecker() *SecurityChecker {
	return &SecurityChecker{
		sensitivePatterns: map[string][]string{
			"file_access": {
				`os\.(Open|Create|Remove|RemoveAll|Chmod|Chown)`,
				`ioutil\.(ReadFile|WriteFile)`,
			},
			"network_access": {
				`net\.(Dial|Listen)`,
				`http\.(Get|Post|Put|Delete)`,
			},
			"code_execution": {
				`exec\.(Command|Run)`,
				`syscall\.(Exec|StartProcess)`,
			},
			"input_validation": {
				`fmt\.(Scan|Scanf|Scanln)`,
				`bufio\.NewScanner`,
			},
			"random_generation": {
				`math/rand\.(Int|Float|Perm)`,
				`crypto/rand\.(Read|Prime)`,
			},
			"sensitive_data": {
				`(?i)(password|secret|key|token|credential)`,
				`fmt\.Printf.*password`,
			},
		},
	}
}

// CheckMemoryUsage 检查内存使用情况
func (c *SecurityChecker) CheckMemoryUsage(filePath string) (uint64, error) {
	var m runtime.MemStats
	runtime.ReadMemStats(&m)
	initialAlloc := m.Alloc

	// 读取并执行文件
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return 0, fmt.Errorf("读取文件失败: %v", err)
	}

	// 解析文件以检查内存使用
	fset := token.NewFileSet()
	_, err = parser.ParseFile(fset, filePath, content, parser.AllErrors)
	if err != nil {
		return 0, fmt.Errorf("解析文件失败: %v", err)
	}

	runtime.ReadMemStats(&m)
	finalAlloc := m.Alloc

	return finalAlloc - initialAlloc, nil
}

// CheckExecutionTime 检查执行时间
func (c *SecurityChecker) CheckExecutionTime(filePath string, timeout time.Duration) error {
	done := make(chan bool)
	var execErr error

	go func() {
		// 读取并解析文件
		content, err := ioutil.ReadFile(filePath)
		if err != nil {
			execErr = fmt.Errorf("读取文件失败: %v", err)
			done <- true
			return
		}

		fset := token.NewFileSet()
		_, err = parser.ParseFile(fset, filePath, content, parser.AllErrors)
		if err != nil {
			execErr = fmt.Errorf("解析文件失败: %v", err)
			done <- true
			return
		}

		done <- true
	}()

	select {
	case <-done:
		return execErr
	case <-time.After(timeout):
		return fmt.Errorf("执行超时(>%v)", timeout)
	}
}

// CheckFileAccess 检查文件访问安全性
func (c *SecurityChecker) CheckFileAccess(filePath string) ([]string, error) {
	violations := make([]string, 0)
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("读取文件失败: %v", err)
	}

	c.mu.RLock()
	patterns := c.sensitivePatterns["file_access"]
	c.mu.RUnlock()

	for _, pattern := range patterns {
		re, err := regexp.Compile(pattern)
		if err != nil {
			continue
		}

		matches := re.FindAllString(string(content), -1)
		for _, match := range matches {
			violations = append(violations, fmt.Sprintf("发现敏感文件操作: %s", match))
		}
	}

	return violations, nil
}

// CheckNetworkAccess 检查网络访问安全性
func (c *SecurityChecker) CheckNetworkAccess(filePath string) ([]string, error) {
	violations := make([]string, 0)
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("读取文件失败: %v", err)
	}

	c.mu.RLock()
	patterns := c.sensitivePatterns["network_access"]
	c.mu.RUnlock()

	for _, pattern := range patterns {
		re, err := regexp.Compile(pattern)
		if err != nil {
			continue
		}

		matches := re.FindAllString(string(content), -1)
		for _, match := range matches {
			violations = append(violations, fmt.Sprintf("发现敏感网络操作: %s", match))
		}
	}

	return violations, nil
}

// CheckInputValidation 检查输入验证
func (c *SecurityChecker) CheckInputValidation(filePath string) ([]string, error) {
	issues := make([]string, 0)
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("读取文件失败: %v", err)
	}

	fset := token.NewFileSet()
	file, err := parser.ParseFile(fset, filePath, content, parser.AllErrors)
	if err != nil {
		return nil, fmt.Errorf("解析文件失败: %v", err)
	}

	ast.Inspect(file, func(n ast.Node) bool {
		if call, ok := n.(*ast.CallExpr); ok {
			if sel, ok := call.Fun.(*ast.SelectorExpr); ok {
				if x, ok := sel.X.(*ast.Ident); ok {
					funcName := x.Name + "." + sel.Sel.Name
					if strings.Contains(funcName, "fmt.Scan") || strings.Contains(funcName, "bufio.NewScanner") {
						issues = append(issues, fmt.Sprintf("未验证的输入: %s", funcName))
					}
				}
			}
		}
		return true
	})

	return issues, nil
}

// CheckRandomGeneration 检查随机数生成安全性
func (c *SecurityChecker) CheckRandomGeneration(filePath string) ([]string, error) {
	issues := make([]string, 0)
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("读取文件失败: %v", err)
	}

	c.mu.RLock()
	patterns := c.sensitivePatterns["random_generation"]
	c.mu.RUnlock()

	for _, pattern := range patterns {
		re, err := regexp.Compile(pattern)
		if err != nil {
			continue
		}

		matches := re.FindAllString(string(content), -1)
		for _, match := range matches {
			if !strings.Contains(match, "crypto/rand") {
				issues = append(issues, fmt.Sprintf("不安全的随机数生成: %s", match))
			}
		}
	}

	return issues, nil
}

// CheckSensitiveData 检查敏感数据处理
func (c *SecurityChecker) CheckSensitiveData(filePath string) ([]string, error) {
	issues := make([]string, 0)
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("读取文件失败: %v", err)
	}

	c.mu.RLock()
	patterns := c.sensitivePatterns["sensitive_data"]
	c.mu.RUnlock()

	for _, pattern := range patterns {
		re, err := regexp.Compile(pattern)
		if err != nil {
			continue
		}

		matches := re.FindAllString(string(content), -1)
		for _, match := range matches {
			issues = append(issues, fmt.Sprintf("敏感数据泄露风险: %s", match))
		}
	}

	return issues, nil
}

// CheckSandboxEscape 检查沙箱逃逸
func (c *SecurityChecker) CheckSandboxEscape(filePath string) ([]string, error) {
	violations := make([]string, 0)
	content, err := ioutil.ReadFile(filePath)
	if err != nil {
		return nil, fmt.Errorf("读取文件失败: %v", err)
	}

	fset := token.NewFileSet()
	file, err := parser.ParseFile(fset, filePath, content, parser.AllErrors)
	if err != nil {
		return nil, fmt.Errorf("解析文件失败: %v", err)
	}

	ast.Inspect(file, func(n ast.Node) bool {
		if call, ok := n.(*ast.CallExpr); ok {
			if sel, ok := call.Fun.(*ast.SelectorExpr); ok {
				if x, ok := sel.X.(*ast.Ident); ok {
					funcName := x.Name + "." + sel.Sel.Name
					if strings.Contains(funcName, "os.") || strings.Contains(funcName, "exec.") {
						violations = append(violations, fmt.Sprintf("危险的系统调用: %s", funcName))
					}
				}
			}
		}
		return true
	})

	return violations, nil
}

// PerformFullCheck 执行完整的安全检查
func (c *SecurityChecker) PerformFullCheck(filePath string) (map[string]interface{}, error) {
	results := make(map[string]interface{})

	// 检查内存使用
	memoryUsage, err := c.CheckMemoryUsage(filePath)
	if err != nil {
		results["memory_usage"] = err.Error()
	} else {
		results["memory_usage"] = memoryUsage
	}

	// 检查执行时间
	err = c.CheckExecutionTime(filePath, 5*time.Second)
	if err != nil {
		results["execution_time"] = err.Error()
	} else {
		results["execution_time"] = "OK"
	}

	// 检查文件访问
	fileAccess, err := c.CheckFileAccess(filePath)
	if err != nil {
		results["file_access"] = err.Error()
	} else {
		results["file_access"] = fileAccess
	}

	// 检查网络访问
	networkAccess, err := c.CheckNetworkAccess(filePath)
	if err != nil {
		results["network_access"] = err.Error()
	} else {
		results["network_access"] = networkAccess
	}

	// 检查输入验证
	inputValidation, err := c.CheckInputValidation(filePath)
	if err != nil {
		results["input_validation"] = err.Error()
	} else {
		results["input_validation"] = inputValidation
	}

	// 检查随机数生成
	randomGeneration, err := c.CheckRandomGeneration(filePath)
	if err != nil {
		results["random_generation"] = err.Error()
	} else {
		results["random_generation"] = randomGeneration
	}

	// 检查敏感数据
	sensitiveData, err := c.CheckSensitiveData(filePath)
	if err != nil {
		results["sensitive_data"] = err.Error()
	} else {
		results["sensitive_data"] = sensitiveData
	}

	// 检查沙箱逃逸
	sandboxEscape, err := c.CheckSandboxEscape(filePath)
	if err != nil {
		results["sandbox_escape"] = err.Error()
	} else {
		results["sandbox_escape"] = sandboxEscape
	}

	return results, nil
} 

File: movery/utils/security.py
import os
import ast
import time
import psutil
import socket
import threading
from typing import List, Dict, Any, Optional
import re
import astroid

class SecurityChecker:
    """安全检查器类"""
    
    def __init__(self):
        """初始化安全检查器"""
        self.sensitive_patterns = {
            'file_access': [
                r'open\s*\([^)]*[\'"]\/(?:etc|root|home|usr|var)[^\'"]*[\'"]\s*\)',
                r'os\.(?:remove|unlink|rmdir|mkdir|chmod|chown)',
            ],
            'network_access': [
                r'socket\.(?:socket|connect|bind|listen)',
                r'urllib\.(?:request|urlopen)',
                r'requests\.(?:get|post|put|delete)',
            ],
            'code_execution': [
                r'(?:exec|eval|subprocess\.(?:call|Popen|run))',
                r'os\.(?:system|popen|spawn)',
            ],
            'input_validation': [
                r'input\s*\(',
                r'raw_input\s*\(',
                r'eval\s*\(',
            ],
            'random_generation': [
                r'random\.(?:random|randint|choice|randrange)',
                r'secrets\.(?:token_hex|token_bytes|token_urlsafe)',
            ],
            'sensitive_data': [
                r'(?:password|secret|key|token|credential)',
                r'print\s*\([^)]*(?:password|secret|key|token)[^)]*\)',
            ],
        }

    def check_memory_usage(self, file_path: str) -> Dict[str, Any]:
        """检查文件执行时的内存使用情况

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            process = psutil.Process()
            initial_memory = process.memory_info().rss
            
            # 在新线程中执行代码以便监控
            def execute_code():
                with open(file_path, 'r') as f:
                    exec(f.read())
                    
            thread = threading.Thread(target=execute_code)
            thread.start()
            thread.join(timeout=5)  # 最多等待5秒
            
            final_memory = process.memory_info().rss
            memory_usage = final_memory - initial_memory
            
            with open(file_path, 'r') as f:
                content = f.read()
            
            patterns = [r'list\s*\(.*\)', r'\[\s*.*\s*for\s.*\]', r'dict\s*\(.*\)']
            matches = []
            for pattern in patterns:
                matches.extend(re.finditer(pattern, content))
            
            return {
                "has_issues": memory_usage > 100 * 1024 * 1024,  # 超过100MB认为有问题
                "issues": [f"内存使用量过大: {memory_usage / 1024 / 1024:.2f}MB"] if memory_usage > 100 * 1024 * 1024 else [],
                "details": {match.group(): match.start() for match in matches},
                "patterns": patterns
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"内存检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def check_execution_time(self, file_path: str, timeout: float = 5.0) -> Dict[str, Any]:
        """检查文件执行时间

        Args:
            file_path: 待检查的文件路径
            timeout: 超时时间(秒)

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            start_time = time.time()
            
            def execute_code():
                with open(file_path, 'r') as f:
                    exec(f.read())
                    
            thread = threading.Thread(target=execute_code)
            thread.start()
            thread.join(timeout=timeout)
            
            execution_time = time.time() - start_time
            is_timeout = thread.is_alive()
            
            with open(file_path, 'r') as f:
                content = f.read()
            
            patterns = [r'while\s+True:', r'time\.sleep\s*\(', r'for\s+.*\s+in\s+range\s*\(\s*\d+\s*\)']
            matches = []
            for pattern in patterns:
                matches.extend(re.finditer(pattern, content))
            
            return {
                "has_issues": is_timeout or execution_time > timeout,
                "issues": [f"执行超时(>{timeout}秒)"] if is_timeout else [f"执行时间过长: {execution_time:.2f}秒"] if execution_time > timeout else [],
                "details": {match.group(): match.start() for match in matches},
                "patterns": patterns
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"执行时间检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def check_file_access(self, file_path: str) -> Dict[str, Any]:
        """检查文件访问安全性

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            violations = []
            matches_dict = {}
            with open(file_path, 'r') as f:
                content = f.read()
                
            for pattern in self.sensitive_patterns['file_access']:
                matches = list(re.finditer(pattern, content))
                for match in matches:
                    violations.append(f"发现敏感文件操作: {match.group()}")
                    matches_dict[match.group()] = match.start()
                    
            return {
                "has_issues": len(violations) > 0,
                "issues": violations,
                "details": matches_dict,
                "patterns": self.sensitive_patterns['file_access']
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"文件访问检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def check_network_access(self, file_path: str) -> Dict[str, Any]:
        """检查网络访问安全性

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            violations = []
            matches_dict = {}
            with open(file_path, 'r') as f:
                content = f.read()
                
            for pattern in self.sensitive_patterns['network_access']:
                matches = list(re.finditer(pattern, content))
                for match in matches:
                    violations.append(f"发现敏感网络操作: {match.group()}")
                    matches_dict[match.group()] = match.start()
                    
            return {
                "has_issues": len(violations) > 0,
                "issues": violations,
                "details": matches_dict,
                "patterns": self.sensitive_patterns['network_access']
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"网络访问检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def check_input_validation(self, file_path: str) -> Dict[str, Any]:
        """检查输入验证

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            issues = []
            matches_dict = {}
            with open(file_path, 'r') as f:
                content = f.read()
                
            try:
                module = ast.parse(content)
                for node in ast.walk(module):
                    if isinstance(node, ast.Call):
                        if isinstance(node.func, ast.Name):
                            func_name = node.func.id
                            if func_name in ['input', 'raw_input']:
                                issues.append(f"未验证的输入: 第{node.lineno}行")
                                matches_dict[func_name] = node.lineno
                        elif isinstance(node.func, ast.Attribute):
                            if node.func.attr in ['get', 'post', 'put', 'delete']:
                                issues.append(f"未验证的HTTP请求: 第{node.lineno}行")
                                matches_dict[f"{node.func.value.id}.{node.func.attr}"] = node.lineno
            except:
                issues.append("代码解析失败")
                
            return {
                "has_issues": len(issues) > 0,
                "issues": issues,
                "details": matches_dict,
                "patterns": self.sensitive_patterns['input_validation']
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"输入验证检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def check_random_generation(self, file_path: str) -> Dict[str, Any]:
        """检查随机数生成安全性

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            issues = []
            matches_dict = {}
            with open(file_path, 'r') as f:
                content = f.read()
                
            for pattern in self.sensitive_patterns['random_generation']:
                matches = list(re.finditer(pattern, content))
                for match in matches:
                    if 'secrets' not in match.group():
                        issues.append(f"不安全的随机数生成: {match.group()}")
                        matches_dict[match.group()] = match.start()
                    
            return {
                "has_issues": len(issues) > 0,
                "issues": issues,
                "details": matches_dict,
                "patterns": self.sensitive_patterns['random_generation']
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"随机数生成检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def check_sensitive_data(self, file_path: str) -> Dict[str, Any]:
        """检查敏感数据处理

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            issues = []
            matches_dict = {}
            with open(file_path, 'r') as f:
                content = f.read()
                
            for pattern in self.sensitive_patterns['sensitive_data']:
                matches = list(re.finditer(pattern, content))
                for match in matches:
                    issues.append(f"敏感数据泄露风险: {match.group()}")
                    matches_dict[match.group()] = match.start()
                    
            return {
                "has_issues": len(issues) > 0,
                "issues": issues,
                "details": matches_dict,
                "patterns": self.sensitive_patterns['sensitive_data']
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"敏感数据检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def check_sandbox_escape(self, file_path: str) -> Dict[str, Any]:
        """检查沙箱逃逸

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        try:
            violations = []
            matches_dict = {}
            with open(file_path, 'r') as f:
                content = f.read()
                
            try:
                module = astroid.parse(content)
                for node in module.nodes_of_class(astroid.Call):
                    if isinstance(node.func, astroid.Attribute):
                        if node.func.attrname in ['system', 'popen', 'spawn', 'call', 'Popen', 'run']:
                            violations.append(f"危险的系统调用: {node.as_string()}")
                            matches_dict[node.as_string()] = node.lineno
                    elif isinstance(node.func, astroid.Name):
                        if node.func.name in ['exec', 'eval']:
                            violations.append(f"危险的代码执行: {node.as_string()}")
                            matches_dict[node.as_string()] = node.lineno
            except:
                violations.append("代码解析失败")
                
            return {
                "has_issues": len(violations) > 0,
                "issues": violations,
                "details": matches_dict,
                "patterns": self.sensitive_patterns['code_execution']
            }
        except Exception as e:
            return {
                "has_issues": True,
                "issues": [f"沙箱逃逸检查失败: {str(e)}"],
                "details": {},
                "patterns": []
            }

    def perform_full_check(self, file_path: str) -> Dict[str, Any]:
        """执行完整的安全检查

        Args:
            file_path: 待检查的文件路径

        Returns:
            Dict[str, Any]: 检查结果
        """
        results = {
            'memory_usage': self.check_memory_usage(file_path),
            'execution_time': self.check_execution_time(file_path),
            'file_access': self.check_file_access(file_path),
            'network_access': self.check_network_access(file_path),
            'input_validation': self.check_input_validation(file_path),
            'random_generation': self.check_random_generation(file_path),
            'sensitive_data': self.check_sensitive_data(file_path),
            'sandbox_escape': self.check_sandbox_escape(file_path)
        }
        
        return results
